File: go/challenge01/valid_password.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
)

const (
	StateStart = iota
	StateDigit
	StateChar
)

func isValid(pwd string) bool {
	state := StateStart
	lastDigit := '0'
	lastChar := 'a'

	for _, c := range pwd {
		switch {
		case c >= '0' && c <= '9':
			if state == StateChar {
				return false
			}
			state = StateDigit
			if c < lastDigit {
				return false
			}
			lastDigit = c
		case c >= 'a' && c <= 'z':
			state = StateChar
			if c < lastChar {
				return false
			}
			lastChar = c
		default:
			return false
		}
	}
	return true
}

func main() {
	file, err := os.Open("log.txt")
	if err != nil {
		log.Fatalf("Error opening file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	valid, invalid := 0, 0

	for scanner.Scan() {
		pwd := strings.TrimSpace(scanner.Text())
		if pwd == "" {
			continue
		}
		if isValid(pwd) {
			valid++
		} else {
			invalid++
		}
	}
	if err := scanner.Err(); err != nil {
		log.Fatalf("Error reading lines: %v", err)
	}
	fmt.Printf("submit %dtrue%dfalse\n", valid, invalid)
}

File: node/challenge01/valid_password.js
const fs = require('fs');
const readline = require('readline');

function isValid(pwd) {
    let state = 'start';
    let lastDigit = '0';
    let lastChar = 'a';

    for (const c of pwd) {
        if (c >= '0' && c <= '9') {
            if (state === 'char') {
                return false;
            }
            state = 'digit';
            if (c < lastDigit) {
                return false;
            }
            lastDigit = c;
        }
        else if (c >= 'a' && c <= 'z') {
            state = 'char';
            if (c < lastChar) {
                return false;
            }
            lastChar = c;
        }
        else {
            return false;
       }
    }
    return true;
}

async function main() {
    const fileStream = fs.createReadStream('log.txt', 'utf8');

    const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
    });

    let valid = 0;
    let invalid = 0;

    for await (const line of rl) {
        const pwd = line.trim();
        if (pwd === '') continue;

        if (isValid(pwd)) valid++;
        else invalid++;
    }
    console.log(`submit ${valid}true${invalid}false`);
}

main().catch(err => {
    console.error('Error:', err);
    process.exit(1);
});